Optimize cagg refresh for small invalidations

erimatnor opened the pull request:
This PR includes two changes:

**Change 1:**
 
The refreshing of a continuous aggregate is slow when many small
invalidations are generated by frequent single row insert
backfills. This change adds an optimization that merges small
invalidations by first expanding invalidations to full bucket
boundaries. There is really no reason to maintain invalidations that
aren't covering full buckets since refresh windows are already aligned
to buckets anyway.

**Change 2:**

When there are many small (e.g., single timestamp) invalidations that
cannot be merged despite expanding invalidations to full buckets
(e.g., invalidations are spread across every second bucket in the
worst case), it might no longer be beneficial to materialize every
invalidation separately.

Instead, this change adds a threshold for the number of invalidations
used by the refresh (currently 10 by default) above which
invalidations are merged into one range based on the lowest and
greatest invalidated time value.

The limit can be controlled by an anonymous session variable for
debugging and tweaking purposes. It might be considered for promotion
to an official GUC in the future.

Fixes #2867

erimatnor commented on tsl/src/continuous_aggs/refresh.c:
```diff
@@ -290,15 +291,71 @@ log_refresh_window(int elevel, const ContinuousAgg *cagg, const InternalTimeRang
 		 DatumGetCString(OidFunctionCall1(outfuncid, end_ts)));
 }
 
+#define MAX_INVALIDATIONS_PER_REFRESH 10
```
The semantics here are different from `max_interval_per_job`. The old `max_interval_per_job` setting is actually already similar to the user- or policy-specified refresh window; i.e., user controls the max amount materialized by specifying a specific refresh window size.

This max value is about how we optimize the materialization of the specified refresh window. Pre-2.0, everything from the earliest invalidation and forward was re-materialized no matter whether the entire range needed re-materialization or not. Post-2.0 we break that down so that we only materialize the regions/buckets that are actually invalid. However, if there are a lot of small regions that are invalidated (e.g., every other bucket in the worst case), it is cheaper to merge that into a large range. The session variable below controls the threshold.

mkindahl commented on tsl/src/continuous_aggs/invalidation.c:
```diff
@@ -458,6 +462,53 @@ invalidation_entry_reset(Invalidation *entry)
+	max_bucket_end = ts_time_bucket_by_type(bucket_width, ts_time_get_max(timetype), timetype);
+	max_bucket_end = ts_time_saturating_sub(max_bucket_end, 1, timetype);
+
+	if (inv->lowest_modified_value < min_bucket_start)
+		inv->lowest_modified_value = PG_INT64_MIN;
+	else if (inv->lowest_modified_value > max_bucket_end)
+		inv->lowest_modified_value = PG_INT64_MAX;
```
Isn't this a problem if `PG_INT16_MIN` or `PG_INT64_MAX` happen to be in the middle of a bucket?

mkindahl commented on tsl/src/continuous_aggs/refresh.c:
```diff
@@ -290,15 +291,71 @@ log_refresh_window(int elevel, const ContinuousAgg *cagg, const InternalTimeRang
+	{
+		char *endptr = NULL;
+
+		/* Not using pg_strtol here since we don't want to throw error in case
+		 * of parsing issue */
+		max_invalidations = strtol(max_invalidations_setting, &endptr, 10);
```
Why not? If the variable is set to a strange string, isn't it better to raise an error early rather than running with a default value? A user making a subtle mistake might be befuddled about why things are behaving strange.

pmwkaa replied:
Can this be checked by the GUC callback or something like that? If there is none, we should validate it

pmwkaa replied:
Sorry, this was not clear, that we are not going to document it

erimatnor replied:
Sorry it wasn't clear. Tried to make clear in the function: "* Use a session variable for now, for debugging and tweaking. Might promote to full GUC if useful."

mkindahl replied:
The point of raising an error here is to ensure that the value is corrected rather than using a default value, so yes, it should throw an error until the value is changed. Validating it when setting it is (of course) better.

erimatnor replied:
Added a warning as agreed.

mkindahl commented on tsl/src/continuous_aggs/refresh.c:
```diff
@@ -290,15 +291,71 @@ log_refresh_window(int elevel, const ContinuousAgg *cagg, const InternalTimeRang
+		 * of parsing issue */
+		max_invalidations = strtol(max_invalidations_setting, &endptr, 10);
+
+		if (*endptr != '\0')
+			max_invalidations = MAX_INVALIDATIONS_PER_REFRESH;
```
What about trailing spaces? Shouldn't that be handled? Also, AFAICT, the empty string will result in 0 max invalidation, which might not want be what you want.

erimatnor replied:
I can add handling for trailing spaces, but I don't think this is necessary for something that is not an official GUC. Falling back to the default value or even 0 is OK. All of these will work.

pmwkaa replied:
Do we need to check `strtol` for overflows as well? errno `ERANGE` or endptr is enought?

erimatnor replied:
In my understanding, checking endptr is enough unless we want to understand more about what failed.

erimatnor replied:
FWIW, if the integer value is too big it will just return LONG_MAX. Added test.

mkindahl commented on tsl/src/continuous_aggs/refresh.c:
```diff
@@ -326,8 +383,39 @@ continuous_agg_refresh_with_window(const ContinuousAgg *cagg,
+		Assert(merged_refresh_window.start >= refresh_window->start);
+		Assert(merged_refresh_window.end <= refresh_window->end);
+
+		log_refresh_window(DEBUG1,
+						   cagg,
+						   &merged_refresh_window,
+						   psprintf("merged %ld invalidations for refresh on", count));
```
I think this should be a notice (https://www.postgresql.org/docs/12/runtime-config-logging.html#RUNTIME-CONFIG-SEVERITY-LEVELS):

> NOTICE | Provides information that might be helpful to users, e.g., notice of truncation of long identifiers.

erimatnor replied:
I disagree. A notice is sent to the client and this is not normally valuable information to the user since it is implementation detail (debug information). Users care that the result of the refresh is what they expect, not _how_ it was executed. If they want to see this info, they can always change the debug level.

mkindahl replied:
The example given in the manual is where something happened that might change the behavior but not likely affect the result, e.g., truncation of identifier. The PostgreSQL code uses notices internally for these things (and others, but these are examples):

- Operations are skipped for statements, e.g., table is not dropped in a drop statement
- Constraints are merged with inherited definition
- Merging multiple inherited definitions

I think this message belongs in the same category.

erimatnor replied:
This is about how we internally process invalidations. Users might not even know what an invalidation is, and most won't care.

mkindahl reviewed (commented):
Found a few things that are not clear, could be, or is a problem. If you do not think they are, please respond to them.

I also think that is not a good idea to have the `max_invalidations_per_refresh` for the user function and especially having it to default to 10 (which it does, from my understanding). For the policy runs, this is just a way to throttle the background workers, which is fine, but when calling this from the command line the result will be that the entire range is not refreshed and the user will have to type the command several times, which will be annoying. I think it might be better to add a parameter to the function which defaults to do a full refresh and ensure that the policy set a value for this parameter instead. This will make sure that policies can be throttled and have a good default value, while command-line calls will refresh the entire range.

pmwkaa commented on tsl/test/sql/continuous_aggs_invalidation.sql:
```diff
@@ -619,3 +619,49 @@ INSERT INTO conditions VALUES(6, 1, 1.0);
+RESET client_min_messages;
+
+-- Allow at most 5 individual invalidations per refreshe
+SET timescaledb.max_invalidations_per_refresh=5;
```
Suggest to add test, that checks this value is correct and produce an error if not

erimatnor replied:
I can do that if deemed important. But I don't think we should waste time on that for something that is just a session variable for debugging and not an real GUC. Also, it is not possible to parse session variables when set, so the failure would appear when running refresh and persist until the variable is changed.

pmwkaa commented on tsl/src/continuous_aggs/invalidation.c:
```diff
@@ -458,6 +462,53 @@ invalidation_entry_reset(Invalidation *entry)
+	max_bucket_end = ts_time_saturating_sub(max_bucket_end, 1, timetype);
+
+	if (inv->lowest_modified_value < min_bucket_start)
+		inv->lowest_modified_value = PG_INT64_MIN;
```
Do we have tests somewhere which use those MIN/MAX as a values?

erimatnor replied:
I am actually unsure what is asked. MIN/MAX is used to mean invalid to -infinity and +infinity. There are several tests (e.g., `continuous_aggs_invalidation`) which show such invalidations in the invalidation log.

pmwkaa reviewed (approved):
Looks good overall, don't have much to add, just a couple of comments

erimatnor commented:
> I also think that is not a good idea to have the `max_invalidations_per_refresh` for the user function and especially having it to default to 10 (which it does, from my understanding). For the policy runs, this is just a way to throttle the background workers, which is fine, but when calling this from the command line the result will be that the entire range is not refreshed and the user will have to type the command several times, which will be annoying. 

So, this seems to be a misunderstanding of what the changes do, and that's on me for failing to make clear and perhaps a badly named variable.

Two things that I'd like to clarify:

* The changes in this PR in no way changes the semantics of `refresh_continuous_aggregate`. After a call to this function, everything in the given window will be up-to-date. This hasn't changed. What happens internally, though, is that we break a refresh down into a set of multiple sub-refreshes on the ranges (buckets) that are actually out-of-date. But if that turns out to be 100 separate refreshes, this is likely to be slower than one single refresh across the range determined by the min invalidated time and max invalidated time. The `max_invalidations_per_refresh` just sets the threshold of where we merge those separate ranges into one big range. It does _not_ mean that we leave some ranges out-of-date.
* None of the changes in this PR has anything to do with throttling policies. Not sure where that impression came from.

erimatnor commented on tsl/src/continuous_aggs/invalidation.c:
```diff
@@ -458,6 +462,53 @@ invalidation_entry_reset(Invalidation *entry)
+	 * and move one step down to the end value of the previous
+	 * bucket. Remember that invalidations are inclusive, so the "greatest"
+	 * value should be the last value of the last full bucket. */
+	max_bucket_end = ts_time_bucket_by_type(bucket_width, ts_time_get_max(timetype), timetype);
```
I thought about it, but figured it wasn't worth it. Perhaps you have a suggestion of how to easily cache it without making this unnecessarily complex?

mkindahl commented on tsl/src/continuous_aggs/invalidation.c:
```diff
@@ -458,6 +462,53 @@ invalidation_entry_reset(Invalidation *entry)
+	int64 min_bucket_start;
+	int64 max_bucket_end;
+
+	/* If the min value falls on the exact start of a full bucket we are
+	 * good. Otherwise, we need to move to the next full bucket. */
+	min_bucket_start =
+		ts_time_saturating_add(ts_time_get_min(timetype), bucket_width - 1, timetype);
+	min_bucket_start = ts_time_bucket_by_type(bucket_width, min_bucket_start, timetype);
```
The comment was hard to read, you might want to clarify it.

erimatnor replied:
which comment are you referring to? Is it the one at line 478 or the comment on the function? How can it be clarified?

mkindahl replied:
It is the comment on line 478. AFAICT, you're computing the beginning of the first bucket completely inside the range of the type, but this is not clear from the comment.

erimatnor replied:
Made an attempt to clarify.

mkindahl replied:
LGTM. Good that you did the same change for the max variable.

mkindahl commented:
> > I also think that is not a good idea to have the `max_invalidations_per_refresh` for the user function and especially having it to default to 10 (which it does, from my understanding). For the policy runs, this is just a way to throttle the background workers, which is fine, but when calling this from the command line the result will be that the entire range is not refreshed and the user will have to type the command several times, which will be annoying.
> 
> So, this seems to be a misunderstanding of what the changes do, and that's on me for failing to make clear and perhaps a badly named variable.
> 
> Two things that I'd like to clarify:
> 
>     * The changes in this PR in no way changes the semantics of `refresh_continuous_aggregate`. After a call to this function, everything in the given window will be up-to-date. This hasn't changed. What happens internally, though, is that we break a refresh down into a set of multiple sub-refreshes (materialization) on the ranges (buckets) that are actually out-of-date. But if that turns out to be 100 separate refreshes, this is likely to be slower than one single refresh across the range determined by the min invalidated time and max invalidated time. The `max_invalidations_per_refresh` just sets the threshold of where we merge those separate ranges into one big range. It does _not_ mean that we leave some ranges out-of-date.
> 
>     * None of the changes in this PR has anything to do with throttling policies. Not sure where that impression came from.
> 
> 
> I've tried to change the comments and variable names to make this more clear.

I understand what the variable do, but IMHO the problem is that the default behavior is to not process the entire range when it is "too complicated" (when we have a lot of buckets that are not connected). We had several comments before when `REFRESH MATERIALIZED VIEW` didn't refresh everything, so I think we should make sure that we, by default, refresh the entire range even if that happens to take a long time.

erimatnor commented:
> I understand what the variable do, but IMHO the problem is that the default behavior is to not process the entire range when it is "too complicated" (when we have a lot of buckets that are not connected). We had several comments before when `REFRESH MATERIALIZED VIEW` didn't refresh everything, so I think we should make sure that we, by default, refresh the entire range even if that happens to take a long time.

Above the threshold we actually process more than necessary, so it is actually the opposite of your understanding. When you say we had comments about `REFRESH MATERIALIZED VIEW`, I assume you are referring to pre-2.0 behavior. That functionality was entirely different from what we do here, so there is no clear comparison. The semantics of `refresh_continuous_aggregate` here is to leave everything in the window/range up-to-date after the call completes. So, your statement about not processing the entire range is incorrect.

mkindahl reviewed (approved):
Thanks for the clarification that the changes does not affect the range that the refresh function process. LGTM.

erimatnor commented:
@mkindahl @pmwkaa Changed the following based on feedback.

* Updated comments for min/max bucket calculation (also made a small tweak to calculating the end of the last bucket)
* Added warning for the session variable parsing, and also added support for trailing whitespaces
* Added tests for non-parsable input of session variable

mkindahl commented:
> > I understand what the variable do, but IMHO the problem is that the default behavior is to not process the entire range when it is "too complicated" (when we have a lot of buckets that are not connected). We had several comments before when `REFRESH MATERIALIZED VIEW` didn't refresh everything, so I think we should make sure that we, by default, refresh the entire range even if that happens to take a long time.
> 
> Above the threshold we actually process more than necessary, so it is actually the opposite of your understanding. When you say we had comments about `REFRESH MATERIALIZED VIEW`, I assume you are referring to pre-2.0 behavior. That functionality was entirely different from what we do here, so there is no clear comparison. The semantics of `refresh_continuous_aggregate` here is to leave everything in the user-specified window/range up-to-date after the call completes. So, your statement about not processing the entire range is not the way it works.

Good to hear. Then I misunderstood how it worked.
